crates/nu-cli/src/completions/file_completions.rs

```diff
@@ -157,3 +151,30 @@ pub fn matches(partial: &str, from: &str) -> bool {
     from.to_ascii_lowercase()
         .starts_with(&partial.to_ascii_lowercase())
 }
+
+// escape paths that contains some special characters
+pub fn escape_path_str(path: String) -> String {
+    let mut path = path;
+
+    // Check if path needs to be escaped
+    let needs_escape = path.bytes().fold(
+        false,
+        |acc, x| {
+            acc
+        || (x >> 4) == 0b0010
+        || x == b'\\' // 0x5c
+        || x == b'`'
+        }, // 0x60
+    );
+
+    if needs_escape {
+        // Escape characters
+        path = path.replace('\\', "\\\\");
+        path = path.replace('"', "\\\"");
+
+        // Wrap with double quotes
+        path = format!("\"{}\"", path);
```

Comment:
This can be done similarly with a single pass instead of iterating over the entire string twice and jumping into the massive machinery of format!, since we still aren't escaping any multibyte characters:

```rust
let special_characters = b"\\\"" // operate on bytes since all special characters are ASCII
let replacements = [b"\\\\", b"\\\""];
let mut result: String;
if needs_escape {
    result = String::with_capacity(path.len()*2 + 2); // worst case, escape everything
    result.push(b"\"");

    for b in path.bytes() {
        // basically the equivalent of str.find(), but expanded since this is a &[u8] which AFAIK does not have find()   
        if let Some(idx) = special_characters.iter().enumerate().fold(None, |idx, c| if c.1 == b {Some(c.0)} else {idx}) {
            for rb in replacements[idx].bytes() {
                result.push(rb);
            }
        } else {
            result.push(b);
        }
    }
    result.push("\"");
} else {
    result = path;
}
return result;
```

Reply:
This also eliminates 2 out of the 3 String allocations from that part of the code.

Reply:
Updated code to actually use a Vec<u8> for replacements instead of String.

Reply:
I'm not sure if we should go away from the massive machinery of `format!` to the dark path of `unsafe`.
If this were like hundreds of thousands of files I'd say that we need to think. But it's not the case

Reply:
IMO, this is clearly becoming premature optimization.

Reply:
The only thing `format!` is being invoked for is prepending and appending a single quote character to the string, seems pretty unnecessary to me.

As for the unsafe, there is also the safe function `String::from_utf8(...).unwrap()` which does the same thing except it also checks that the string is actually valid UTF-8.